### PR TITLE
[mlir] add option to print SSA IDs using `NameLoc`s as prefixes

### PR DESCRIPTION
**Pull request by makslevental**
This PR adds an `AsmPrinter` option `-mlir-use-nameloc-as-prefix` which uses trailing `NameLoc`s, if the source IR provides them, as prefixes when printing SSA IDs. **Note**: this PR only changes `AsmPrinter`.

For example:

```mlir
%1 = memref.load %0[] : memref<i32> loc("alice")
```

prints

```mlir
%alice = memref.load %0[] : memref<i32>
```

Currently single/multiple results and bb args are handled:

```mlir
%1:2 = call @make_two_results() : () -> (index, index) loc("bar")
%2:2 = scf.while (%arg1 = %1#0, %arg2 = %1#0) : (index, index) -> (index, index) {
  %3 = arith.cmpi slt, %arg1, %arg2 : index
  scf.condition(%3) %arg1, %arg2 : index, index
} do {
^bb0(%arg3 : index loc("alice"), %arg4: index loc("bob")):
  %c1, %c2 = func.call @make_two_results() : () -> (index, index) loc("harriet")
  scf.yield %c2, %c2 : index, index
} loc("kevin")
```

becomes

```mlir
%bar:2 = call @make_two_results() : () -> (index, index)
%kevin:2 = scf.while (%arg1 = %bar#0, %arg2 = %bar#0) : (index, index) -> (index, index) {
  %0 = arith.cmpi slt, %arg1, %arg2 : index
  scf.condition(%0) %arg1, %arg2 : index, index
} do {
^bb0(%alice: index, %bob: index):
  %harriet:2 = func.call @make_two_results() : () -> (index, index)
  scf.yield %harriet#1, %harriet#1 : index, index
}
```

Note, this approach preserves the `NameLoc` through passes so 

```mlir
// -test-loop-unrolling='unroll-factor=2' -mlir-use-nameloc-as-prefix
scf.for %arg2 = %c0 to %c4 step %c1 {
  %0 = memref.load %arg0[%arg2] : memref<4xf32> loc("foo")
  memref.store %0, %arg1[%arg2] : memref<4xf32>
}
```

becomes

```mlir
scf.for %arg2 = %c0 to %c4 step %c2 {
  %foo = memref.load %arg0[%arg2] : memref<4xf32>
  memref.store %foo, %arg1[%arg2] : memref<4xf32>
  ...
  %foo_1 = memref.load %arg0[%1] : memref<4xf32>
  memref.store %foo_1, %arg1[%1] : memref<4xf32>
}
```

### Call outs

The changes here are also compatible with `OpAsmOpInterface`. Note though, if an op implements `getAsmBlockArgumentNames` but not `getAsmResultNames` (like [`linalg.generic`](https://github.com/makslevental/llvm-project/blob/288f05f63e5f3246657aca9561d75b2aa02cb6f5/mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td#L56)) then the affixed `loc("...")` will not be processed because [`setResultNameFn`](https://github.com/llvm/llvm-project/blob/d61562c967b3a6f0c34e00c90155d16580288d8a/mlir/lib/IR/AsmPrinter.cpp#L1601) is never called.

### Testing

Besides the added lit test, I "stress tested" this by turning it on by default and checking if anything broke ([this commit](https://github.com/llvm/llvm-project/pull/119996/commits/bbca902568bfb047ef67a96dad631314669fd6f9)), then fixing my bugs and stress testing again (this [commit](https://github.com/llvm/llvm-project/pull/119996/commits/ca084f95b2c1a424392049bc4a8555f4f23d83fe) and this [passing test](https://buildkite.com/llvm-project/github-pull-requests/builds/129062#0193c8fd-9284-4c5f-b4c0-ec337d75f7be), Windows test fail there was a flake).

### CODE REVIEW

File: mlir/lib/IR/AsmPrinter.cpp [hunk 1]
```diff
@@ -195,6 +196,10 @@ struct AsmPrinterOptions {
       "mlir-print-unique-ssa-ids", llvm::cl::init(false),
       llvm::cl::desc("Print unique SSA ID numbers for values, block arguments "
                      "and naming conflicts across all regions")};
+
+  llvm::cl::opt<bool> useNameLocAsPrefix{
+      "mlir-use-nameloc-as-prefix", llvm::cl::init(false),
+      llvm::cl::desc("Print SSA IDs using NameLocs as prefixes")};
```
Comment (jpienaar) [inline mlir/lib/IR/AsmPrinter.cpp:202]: Is this all or only those not set via OpAsmInterface?

**Review by jpienaar — COMMENTED** [review]
Overall I think this is nice, minimal start. Could this be documented somewhere? (Potentially debugging guide could be good one).

File: mlir/lib/IR/AsmPrinter.cpp (hunk 1 above)
Comment (makslevental) [reply mlir/lib/IR/AsmPrinter.cpp:202]: It's all - this actually takes precedence over OpAsmInterface. I figured if you turn this on then you want even more legibility than OpAsmInterface provides.

**Comment by makslevental** [comment]
> Overall I think this is nice, minimal start. Could this be documented somewhere? (Potentially debugging guide could be good one).

Sure I can add some notes there.

File: mlir/lib/IR/AsmPrinter.cpp [hunk 2]
```diff
@@ -1553,7 +1570,12 @@ void SSANameState::numberValuesInBlock(Block &block) {
       specialNameBuffer.resize(strlen("arg"));
       specialName << nextArgumentID++;
     }
-    setValueName(arg, specialName.str());
+    if (printerFlags.shouldUseNameLocAsPrefix() && isa<NameLoc>(arg.getLoc())) {
+      auto nameLoc = cast<NameLoc>(arg.getLoc());
+      setValueName(arg, nameLoc.getName());
```
Comment (River707) [inline mlir/lib/IR/AsmPrinter.cpp:1575]: Can you share the implementation of these three usages? Seems like they all pretty much have the same pattern.

Also for the NameLoc itself, you likely want to use `arg.getLoc().findInstanceOf<NameLoc>()`, given that in some cases (e.g. DebugInfo) you may not have a top level NameLoc.

**Review by Wheest — COMMENTED** [review]
Overall this a great way to incrementally bring us closer to some of the goals set out [here](https://discourse.llvm.org/t/retain-original-identifier-names-for-debugging/76417/49).

I see that multiple results are supported, but  would it make sense here to support have different names for different results, e.g.:

```mlir
%kevin, %alice = call @make_two_results() : () -> (index, index)
```

Or even a mix of result groups:

```mlir
%kevin:2, %alice = call @make_three_results() : () -> (index, index, index)
```

This was supported in the more invasive #79704.

https://github.com/llvm/llvm-project/blob/b257ab53e508e5fb00e16a1c73e7fceeea4d0923/mlir/test/IR/print-retain-identifiers.mlir#L82-L83

I would imagine the syntax being something like:

```mlir
%kevin, %alice = call @make_two_results() : () -> (index, index) loc("kevin", "alice")
%kevin:2, %alice = call @make_three_results() : () -> (index, index, index) loc("kevin", "", "alice") // empty string means use result group
%kevin, %alice:2 = call @make_three_results() : () -> (index, index, index) loc("kevin", "alice") // implicit grouping 
```

Although given this is building off of the existing `NamedLoc` system, then that would require a bigger change to change its syntax.  

Having a delimiter character,  e.g., `loc("kevin🐢alice")`, ` loc("kevin🐢🐢alice")` could be a way to do this, but 🐢 would need to be chosen carefully.

**Comment by makslevental** [comment]
> I see that multiple results are supported, but would it make sense here to support have different names for different results, e.g.:

Ya that's tough exactly because you need to introduce a delimiter and then split on it. And then if something about how results are grouped in the future changes, maybe another delimiter would need to be added to support that change. Ie it would create a tiny little "`NameLoc` results dsl" and that requires and actual parser. So I decided it's easier and more robust to just let the existing disambiguation system do its thing.

File: mlir/lib/IR/AsmPrinter.cpp (hunk 2 above)
Comment (makslevental) [reply mlir/lib/IR/AsmPrinter.cpp]: I tried to factor out the whole conditional but then you end up hiding the `printerFlags.shouldUseNameLocAsPrefix()` in the helper itself. I didn't really like that (would be surprising I think) so I factored out the "get the name from a (possibly) nested `NameLoc`".

**Review by jpienaar — APPROVED** [review]
LG overall, thanks

**Comment by joker-eph** [comment]
I don't have concerns with such a change on the printer side: however I do when it'll come to the parser. Does this change make sense without the parser changes?

**Comment by makslevental** [comment]
> Does this change make sense without the parser changes?

Yes - it's exactly architected in this way to minimize coupling to the parser and be useful on its own such that if no parser changes are made, it's still "good enough".

**Pull request merged**